Modernize codebase

kovdan01 opened the pull request:
- Enhance CMakeLists.txt files.
- Move to Boost Test from Google Test to support pre-C++11 compilers.
- Add more configurations on CI matrix builds.
- Add support for the following types:
  - `std::array<std::byte>`
  - `std::span<char>`
  - `std::span<unsigned char>`
  - `std::span<std::byte>`

I plan to submit one more PR to add full support for `std::span` - a nice non-owning data range wrapper from C++20.

redboltz commented:
I merged #952, So #951 should rebase from the new master. In addition, tests for added adaptors are required. Could you do that?

kovdan01 commented:
Yes, I'll do it today

redboltz commented on include/msgpack/v1/adaptor/cpp20/span.hpp:
```diff
@@ -0,0 +1,115 @@
+
+MSGPACK_ADAPTOR_SPAN_BINARY(char)
+MSGPACK_ADAPTOR_SPAN_BINARY(unsigned char)
+MSGPACK_ADAPTOR_SPAN_BINARY(std::byte)
```
I think that the macros are helping template specialization but are not used from other location.
So they can be `#undef`.

kovdan01 replied:
Thank you, done.

redboltz commented on include/msgpack/v1/vrefbuffer.hpp:
```diff
@@ -26,11 +26,16 @@
 #include <sys/uio.h>
+namespace msgpack {
+typedef ::iovec iovec;
+} // namespace msgpack
```
I think that this modification has already been merged.
I guess that rebase process is not done well.

redboltz commented:
I merged #954 and #955 . They helps detecting UB. Could you rebase again from the new master ?

kovdan01 commented:
While trying to add tests for `std::span`, I have noticed that setting `-std=c++20` breaks some tests build:
- A lot of members of `std::allocator` are deleted since C++20: https://en.cppreference.com/w/cpp/memory/allocator, which causes errors in msgpack's custom `test::allocator`.
- New "spaceship" operator from C++20 (`operator<=>`) makes some comparisons ambiguous. Example error:
  ```
  test/object_with_zone.cpp:1115:46: error: ambiguous overload for ‘operator==’ (operand types are ‘std::enable_if<true, msgpack::v1::type::ext>::type’ {aka ‘msgpack::v1::type::ext’} and ‘msgpack::v1::type::ext_ref’)
  ```

It looks like that adding C++20 support will not be trivial. I'll to do it, but it'll take some time (maybe, a week or more).
Actually, it is not related to adding new adaptors, that's why I think that it'll be better to create a new PR for this. Don't you mind?

redboltz commented:
There is no C++20 complete support (language feature and standard libraries) compiler yet.
https://en.cppreference.com/w/cpp/compiler_support

Some of features are already implemented but from my experience, it is usually unstable. I'd like to support C++20 feature after at least one complete support compiler is released.
If some problem would happen, it is difficult to check the problem is caused by msgpack-c implementation, compiler, and/or standard library implementation.
It's my recommendation. But if some feature (mostly adding simple type adaptor) is simple enough, then I accept adding the adaptor. The compiler update would often cause the error in the future, at that time I expect proactive cooperation to fix the issue.

kovdan01 commented:
You are right - none of the compilers has a complete C++20 support. The problem is that if we don't pass `-std=c++20` flag directly (for example, we pass `-std=c++17`), `<span>` header is not found, and we can't compile tests for it. But if we enable `-std=c++20`, some other tests fail to compile. There are some possible solutions:
1) Set compiler flags for span test separately - let other tests have `-std=c++17`, and span tests have `-std=c++20`. I find that it is quite strange and messy and don't like this variant.
2) Delete all span-related stuff until older codebase is ready to be compiled with `-std=c++20` (in terms of this PR it means that we will leave only `std::array<std::byte>` adaptor). If you do not like the next variant, I can do this and re-submit span- and C++20-related stuff in a separate PR later (when you'll be ready to accept it).
3) Fix C++20-related errors in old code. Actually, I have already done it in a separate branch, and [here](https://github.com/kovdan01/msgpack-c/pull/1) you can see the result. The compiler errors could be divided into 2 types:
   - Allocator-related errors. They happen because some functions and types were removed since C++20 (see, for example, [this](https://en.cppreference.com/w/cpp/memory/allocator)). But they were already deprecated since C++17, and alternatives for them (e.g. [std::allocator_traits](https://en.cppreference.com/w/cpp/memory/allocator_traits)) exist since C++11 and are completely stable. Note that to fix this we do not need to change the library code: all fixes are applied to tests.
   - `operator==`-related errors. C++20 introduces a new three-way comparison operator and a bunch of new rules for comparison operators overaload resolution. Citation from N4868 C++ Standard Draft:
     > For the relational (7.6.9) and three-way comparison (7.6.8) operators, the rewritten candidatesalso include a synthesized candidate, with the order of the two parameters reversed, for eachnon-rewritten candidate for the expressiony <=> x

     Because of synthesized overloads with reversed order of parameters, we got errors like this:
     ```
     test/object_with_zone.cpp:1115:46: error: ambiguous overload for ‘operator==’ (operand types are ‘std::enable_if<true, msgpack::v1::type::ext>::type’ {aka ‘msgpack::v1::type::ext’} and ‘msgpack::v1::type::ext_ref’)
     ```
     They occur because `ext` can be implicitly converted to `ext_ref`, and `ext_ref` can be implicitly converted to `ext`. When we try to compare `ext` and `ext_ref`, compiler has two options: (1) convert `ext_ref` to `ext` and call `ext::operator==(const ext&)`; (2) reverse arguments order, convert `ext` to `ext_ref` and call `ext_ref::operator==(const ext_ref&)`. This behavior is correct and predictable according to the standard. To fix the issue, we can either explicitly define comparison operators between `ext` and `ext_ref` or mark `ext(const ext_ref&)` constructor as explicit.

From my perspective, the 3rd variant is applicable - changes are only related to stable and predictable parts of language and standard library. If you think that it is not - I suggest to use the 2nd option.

redboltz commented:
Sorry for my late reply, I had busy days. 3rd option seems good to me.
Is this PR 3rd approach ?

Anyway, could you update (rebase) the PR ? The CI reports an error and I have fixed it by #957.

kovdan01 commented:
Yes, I have updated the PR (rebased from master & added std::span adaptor tests among with c++20 related fixes), now it is 3rd approach. Now I will focus on enhancing CI scripts, particularly, I will add a build variant with C++20 enabled (of course, I will not delete older variants). Without this enhancement some new code just will not be compiled and tested during CI builds.

redboltz commented:
Thanks. I will wait your update.

kovdan01 commented:
@redboltz Hello, sorry for a long delay. I have added appropriate build configurations on CI, and… accidently did a lot of other changes that I believe are important and helpful for the project. The initial motivation was that I simply did not catch the pattern – why that 10 particular sets of options were considered in CI matrix build, and others were not (see [this](https://github.com/msgpack/msgpack-c/blob/b9381f842e0a3c31defb447d598a8a6cce9b007d/.github/workflows/gha.yml#L115)).

If you think that some changes need to be rejected or should be separated into other PR’s – please let me know. Here is a brief overview of what was done:

- **Enhance CMakeLists.txt files.** This includes:
  - Using `target_` versions of commands instead of the global ones (e.g. `link_libraries`). These global commands are considered legacy and should not be used in CMake 3.0 and above.
  - Delete `include_directories`-like commands where not needed – when you write `target_link_libraries(your_target library::name), essential library include directories will already be used. 
  - Use nice cmake options to avoid compiler-specific stuff, for instance, set C++ standard version just like this: ` SET (CMAKE_CXX_STANDARD 11)`
  - Add an install target for msgpackc-cxx library. Previous library releases supported a “canonical” CMake-way of using it – with `find_package(msgpack REQUIRED)` and `target_link_libraries(target msgpackc-cxx)`. Currently library in `cpp_master` does not support it. Although msgpackc-cxx is a header-only library, it is a good idea to create a CMake package for it. 
  - Other minor fixes.
- **Move to Boost Test from Google Test.** The problem with GTest is that latest version that supports pre-C++11 compilers is 1.8.x (2018), while actual version is 1.11.0. Meanwhile, msgpack supports pre-C++11 compilers, and, in my opinion, it is not cool when you are not able to compile C++03 tests without compiling an outdated version of GTest from sources itself instead of installing a pre-compiled library with your package manager (it’s true for macOS and Windows too, not just for Linux). Boost Test supports old compilers, and msgpack depends on Boost already – that’s why I think that Boost Test is a nice choice.
- **Add more configurations on CI matrix builds.** Needless to explain.
- **Enhance C++ version detection.** Using `__cplusplus` macro on MSVC is not a good idea – it is set to 199711L unless /Zc:__cplusplus is specified (see https://docs.microsoft.com/en-us/cpp/build/reference/zc-cplusplus). Instead, it is recommended to use `_MSVC_LANG`. See `MSGPACK_CPP_VERSION` macro definition in msgpack/cpp_version.hpp header.
- Other minor fixes.

kovdan01 commented:
@redboltz Hello! Please let me know if you can review the changes. The PR is pretty big, but I believe that it brings useful changes. Looking forward for your feedback. Thank you in advance!

redboltz commented:
@kovdan01  Thank you for the great work ! 
I just started a review.

Before I post the review comments, I answer your question.

> I have added appropriate build configurations on CI, and… accidently did a lot of other changes that I believe are important and helpful for the project. The initial motivation was that I simply did not catch the pattern – why that 10 particular sets of options were considered in CI matrix build, and others were not (see this).

It is based on https://en.wikipedia.org/wiki/Orthogonal_array_testing but manually configured. The motivation is how to shorten the test time. The resource (especially macos) is limited. To be honest, the criteria of picking case in ambiguous.
It seems that your PR do the all combination of tests using `matrix`. Testing all combination is ideal if the test time is practical.
I created the new branch based on this PR just only for measure the time.
https://github.com/redboltz/msgpack-c/actions/runs/1106831751
If CI finished in practical time, I will apply your approach.

redboltz commented:
I'm watching https://github.com/redboltz/msgpack-c/actions/runs/1106831751 progress.
macos and windows builds run parallelly. But linux builds not run. Maybe enqueued.
When I checked Github Actions behavior, macos, linux, and windows have separated resorces if I remember correctly.

Why linux build is not started ? It seems that macos and linux share (vm or container) resources. Is this right?
Also I noticed that 5 builds run parallelly on macos.

Maybe we should check up to date Github Actions document.

redboltz commented:
I found this post (3 months ago).
https://stackoverflow.com/questions/67295639/how-many-simultaneous-builds-can-be-run-on-github-actions

Maybe total limit is 20.
In our case, 

||macos|linux|windows|total|
---|---|---|---|---
running|5|0|3|8

My expectation as follows:

||macos|linux|windows|total|
---|---|---|---|---
running|5|12|3|20

You can test on **your** account using the following modification:
https://github.com/redboltz/msgpack-c/commit/3546c3fcf67dbf469b4dfe0fe26a9a4b239f87b5

**IMPORTANT**
I guess that the total build limit might be applied by account. I'm not sure by account or by project (repository).

redboltz commented:
It seems that Windows build "Build dependencies" takes a long time. It is about 30 minutes.

redboltz commented:
The current CI on the PR should be improved. The current total time in unacceptable. So far, only 36 builds are finished (they all macos) and takes 34 minutes. Maybe total time will be several hours. I continue measuring.

## Problem and possible solution candidate
- Linux build doesn't run.
   - Need to fix some how.
- Even if Linux builds works parallelly, macos could be a bottle neck.
   - Give up the all matrix build. It is covered by the Linux build. For macos, pick up typical case some how.
- Windows build dependencies takes a long time.
   - I'm not sure what is happening. Use pre-built library as long as you can do.

redboltz commented:
Expected total duration is less that 30 minutes.
https://github.com/msgpack/msgpack-c/actions/runs/825255984

It is a kind of turn around time that  someone post a PR and got the first response (from the CI).

redboltz commented:
2 hours later, macos buils were finished. linux build started. 
It seems that 17 parallel builds are running.

NOTE: windows 3 builds were all finished.

redboltz commented:
time|macos|linux|windows
---|---|---|---
0:00:00|start (5)|queued|start(3)
0:35:00||queued|finished
2:00:00|finished|start(20)|
5:00:00||finished

macos build spent 2 hours.
linux build spent 3 hours.
windows build spent 35 minutes.

redboltz commented:
The PR linux build has the following matrix:

```yaml
matrix
        arch: [32, 64]
        cxx: [98, 11, 14, 17, 20]
        sanitize: ["no", "undefined"]
        api: [1, 2, 3]
        char_sign: ["unsigned", "signed"]
        x3_parse: ["OFF", "ON"]
```

Is this from https://github.com/msgpack/msgpack-c/blob/b9381f842e0a3c31defb447d598a8a6cce9b007d/.github/workflows/gha.yml#L115 ?

I think that you need to pick up some practical patterns as follows:

|api|cxx|arch|char_sign|sanitize|x3_parse|
---|---|---|---|---|---
0|17|64|signed|undefined|ON
1|20|32|unsigned|no|OFF
3|
...|

https://github.com/msgpack/msgpack-c/blob/b9381f842e0a3c31defb447d598a8a6cce9b007d/.github/workflows/gha.yml#L115  was the result of the pick up. 

I think that at most 5 build for macos, at most 15 builds for linux, and windows 3 builds don't need to update.
Windows build requires shrink dependency build time.

Then, the total CI time could be practical.

kovdan01 commented:
@redboltz Thanks for your reply! I see your point – builds took really long time. Updated the PR and left 12 Linux, 5 macOS and 3 Windows builds.

Regarding that

> Windows build requires shrink dependency build time.

I did not change this part – dependencies are installed via vcpkg as before. Github actions cache is used to cache those dependencies, so builds will be fast after first run (while cache is valid). Anyway, Windows builds are not slower than before at the moment.

redboltz commented:
Thank you for updating the PR. I'm very busy now so I couldn't check the PR soon but I will check it when things have calmed down .

Just a comment for CI running time.

The PR spent 51m 49s. It still long.

See the following results:
https://github.com/msgpack/msgpack-c/actions/runs/845095110
https://github.com/msgpack/msgpack-c/actions/runs/825255984

They are recent merged PR. 
You can see it as follows:
1. Open some PR
2. Click `Checks` TAB
3. Click `CI` in the Left sidebar.
4. You can see Total duration
5. Click linux, macos, or windows folder like icon, then you can see individual time.

You can find the bottle neck.

redboltz commented:
It seems that you updated CMakeLists.txt. 
README.md said that cmake >= 3.0.0. Is this still correct?

redboltz commented:
> Thank you for updating the PR. I'm very busy now so I couldn't check the PR soon but I will check it when things have calmed down .
> 
> Just a comment for CI running time.
> 
> The PR spent 51m 49s. It still long.
> 
> See the following results:
> https://github.com/msgpack/msgpack-c/actions/runs/845095110
> https://github.com/msgpack/msgpack-c/actions/runs/825255984
> 
> They are recent merged PR.
> You can see it as follows:
> 
> 1. Open some PR
> 2. Click `Checks` TAB
> 3. Click `CI` in the Left sidebar.
> 4. You can see Total duration
> 5. Click linux, macos, or windows folder like icon, then you can see individual time.
> 
> You can find the bottle neck.

Could you load balance the parallel CI builds ?
Some of builds require short time, and some of tasks require long time.

I realized that 
https://github.com/msgpack/msgpack-c/blob/be4d971c62798eb59f8455dc77a4529748bcd08f/.github/workflows/gha.yml#L114-L209
is the result of load balancing.

redboltz commented:
I tried to build the branch on my local environment.

```sh
[kondo@archboltz] $ rm -rf build                                                                                                                  [kovdan01-add_span_support][~/work/msgpack-c]
[kondo@archboltz] $ mkdir build                                                                                                                   [kovdan01-add_span_support][~/work/msgpack-c]
[kondo@archboltz] $ cd build                                                                                                                      [kovdan01-add_span_support][~/work/msgpack-c]
[kondo@archboltz] $ cmake ..                                                                                                                [kovdan01-add_span_support][~/work/msgpack-c/build]
-- The CXX compiler identification is GNU 11.1.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Boost: /usr/lib64/cmake/Boost-1.76.0/BoostConfig.cmake (found version "1.76.0")
-- Found Doxygen: /usr/bin/doxygen (found version "1.9.1") found components: doxygen dot
-- Configuring done
-- Generating done
-- Build files have been written to: /home/kondo/work/msgpack-c/build
[kondo@archboltz] $ make                                                                                                                    [kovdan01-add_span_support][~/work/msgpack-c/build]
[kondo@archboltz] $                                                                                                                         [kovdan01-add_span_support][~/work/msgpack-c/build]
```

But nothing happens. On cpp_master, started building after `make`.

redboltz commented:
It seems that you replaced gtest with boost test. It is nice. Great work!

I found the following description in the README.md.
I think that it is still correct but for testing, I guess that boost library build is required.
Could you add something about that to the README.md ?

Dependency
-----
msgpack-c requires [boost library](https://www.boost.org/).
msgpack-c depends on only boost headers. You don't need to link boost libraries.

kovdan01 commented:
> I tried to build the branch on my local environment. But nothing happens.

`MSGPACK_BUILD_TESTS` is `OFF` by default, so to build tests you need to pass `-DMSGPACK_BUILD_TESTS=ON` to cmake configuration command. I can change the default value to `ON` if you want - just thought that when a person wants to install a header-only library, they do not want to build its tests.

> README.md said that cmake >= 3.0.0. Is this still correct?

Currently cmake 3.1.0 or newer is required. Fixed it in readme.

> I think that it is still correct but for testing, I guess that boost library build and link are required.
Could you add something about that to the README.md ?

Updated. Thanks!

> Could you load balance the parallel CI builds ?

Yes, I'll do that and will write about results.

redboltz commented:
> > Could you load balance the parallel CI builds ?
> 
> Yes, I'll do that and will write about results.

Thanks. This is not so creative work but it is needed to get result in practical time.

> By the way, I briefly described the changes in the comment on Jul 1 (maybe you did not notice that). For example, it is described what particularly was changed in cmake files (and not only that).

I confirmed https://github.com/msgpack/msgpack-c/pull/951#issuecomment-872334418 , thanks.

kovdan01 commented:
@redboltz Previously I used both gcc and clang for **every** Linux build. It seems that there was no way to enhance build time while using both compilers, that's why now I use gcc for 6 Linux builds and clang for the remaining 6 Linux builds. Currently the same approach is used in `cpp_master` (only one compiler per build).

It seems that after this fix build time is OK (at least as for me).

redboltz commented on include/msgpack/v1/adaptor/raw.hpp:
```diff
@@ -34,7 +34,7 @@ struct raw_ref {
     bool operator== (const raw_ref& x) const
     {
-        return size == x.size && std::memcmp(ptr, x.ptr, size) == 0;
+        return size == x.size && (size == 0 || std::memcmp(ptr, x.ptr, size) == 0);
```
Why size == 0 is needed?

In https://en.cppreference.com/w/cpp/string/byte/memcmp
> ​0​ if all count bytes of lhs and rhs are equal.



In https://en.cppreference.com/w/c/string/byte/memcmp
> Zero if lhs and rhs compare equal, or if count is zero.

I think that if size is 0 then return value is 0.
C notation is very straightforward. C++ notation is difficult to understand but I think that the meaning is the same.

redboltz replied:
It is related topic that the `std::memcmp()` notation is follows:

https://en.cppreference.com/w/cpp/string/byte/memcpy
> If either dest or src is an invalid or null pointer, the behavior is undefined, even if count is zero.

kovdan01 replied:
In https://en.cppreference.com/w/c/string/byte/memcmp

> The behavior is undefined if access occurs beyond the end of either object pointed to by lhs and rhs. **The behavior is undefined if either lhs or rhs is a null pointer.**

I actually found a configuration when this problem was caught with valgrind. To reproduce the problem, revert lines 37, 47 and 53 in `include/msgpack/v1/adaptor/raw.hpp` to their previous state (without checking size against 0) and run this (assuming current directory is repository root):

```
mkdir build-problem && cd build-problem
cmake -GNinja -DMSGPACK_BUILD_TESTS=ON -DMSGPACK_DEFAULT_API_VERSION=1 -DCMAKE_CXX_FLAGS="-g" ..
cmake --build . --target all
ctest -T memcheck
```

The output will contain:

```
      Start 23: raw
23/31 MemCheck #23: raw ..............................***Failed    1.58 sec

97% tests passed, 1 tests failed out of 31

Total Test time (real) = 177.14 sec

The following tests FAILED:
         23 - raw (Failed)
-- Processing memory checking output:
23/31 MemCheck: #23: raw ..............................   Defects: 1
MemCheck log files can be found here: (<#> corresponds to test number)
/home/kovdan01/Projects/msgpack-c/build-problem/Testing/Temporary/MemoryChecker.<#>.log
Memory checking results:
Uninitialized Memory Conditional - 1
        Errors while running CTest
Output from these tests are in: /home/kovdan01/Projects/msgpack-c/build-problem/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```

MemoryChecker.23.log file contains the following:

```
==15705== Memcheck, a memory error detector
==15705== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==15705== Using Valgrind-3.17.0 and LibVEX; rerun with -h for copyright info
==15705== Command: /home/kovdan01/Projects/msgpack-c/build-problem/test/raw
==15705== Parent PID: 15704
==15705==
==15705== Conditional jump or move depends on uninitialised value(s)
==15705==    at 0x48479D0: bcmp (vg_replace_strmem.c:1129)
==15705==    by 0x12855A: msgpack::v1::type::raw_ref::operator==(msgpack::v1::type::raw_ref const&) const (raw.hpp:37)
==15705==    by 0x11AED7: pack_unpack_8_l::test_method() (raw.cpp:47)
==15705==    by 0x11A7B0: pack_unpack_8_l_invoker() (raw.cpp:33)
==15705==    by 0x12D2BB: boost::detail::function::void_function_invoker0<void (*)(), void>::invoke(boost::detail::function::function_buffer&) (function_template.hpp:117)
==15705==    by 0x48AA66D: ??? (in /usr/lib/libboost_unit_test_framework.so.1.76.0)
==15705==    by 0x48A8F5C: boost::execution_monitor::catch_signals(boost::function<int ()> const&) (in /usr/lib/libboost_unit_test_framework.so.1.76.0)
==15705==    by 0x48A8FE6: boost::execution_monitor::execute(boost::function<int ()> const&) (in /usr/lib/libboost_unit_test_framework.so.1.76.0)
==15705==    by 0x48A9098: boost::execution_monitor::vexecute(boost::function<void ()> const&) (in /usr/lib/libboost_unit_test_framework.so.1.76.0)
==15705==    by 0x48D78D6: boost::unit_test::unit_test_monitor_t::execute_and_translate(boost::function<void ()> const&, unsigned long) (in /usr/lib/libboost_unit_test_framework.so.1.76.0)
==15705==    by 0x48AF513: ??? (in /usr/lib/libboost_unit_test_framework.so.1.76.0)
==15705==    by 0x48AF7D7: ??? (in /usr/lib/libboost_unit_test_framework.so.1.76.0)
==15705==
==15705==
==15705== HEAP SUMMARY:
==15705==     in use at exit: 0 bytes in 0 blocks
==15705==   total heap usage: 3,372 allocs, 3,372 frees, 2,271,883 bytes allocated
==15705==
==15705== All heap blocks were freed -- no leaks are possible
==15705==
==15705== Use --track-origins=yes to see where uninitialised values come from
==15705== For lists of detected and suppressed errors, rerun with: -s
==15705== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

So I think that checking size against zero is required, otherwise it leads to UB when one of the pointers is nullptr.

redboltz replied:
I think that it didn't happen before.
I wrote the following simple code.

```cpp
#include <cassert>
#include <cstring>

int main() {
    char* p1 = nullptr;
    char* p2 = nullptr;
    volatile std::size_t n = 0;
    assert(std::memcmp(p1, p2, n) == 0);
}
```

valgrind doesn't report error at least no option.

```
> clang++ -O3 memcmp.cpp && valgrind ./a.out

==135996== Memcheck, a memory error detector
==135996== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==135996== Using Valgrind-3.17.0 and LibVEX; rerun with -h for copyright info
==135996== Command: ./a.out
==135996==
==135996==
==135996== HEAP SUMMARY:
==135996==     in use at exit: 0 bytes in 0 blocks
==135996==   total heap usage: 1 allocs, 1 frees, 72,704 bytes allocated
==135996==
==135996== All heap blocks were freed -- no leaks are possible
==135996==
==135996== For lists of detected and suppressed errors, rerun with: -s
==135996== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

Could you tell me valgrined option and write a minimal code that reproduces the valgrined error?
I want to know why it happens.

redboltz replied:
I removed size == 0 checking from raw.hpp and build on local environment and execute with valgrined. No errors are reported.
I use `valgrind-3.17.0` with no option.

kovdan01 replied:
> Could you tell me valgrined option and write a minimal code that reproduces the valgrined error?
I want to know why it happens.

It seems that valgrind does not detect problems in your code sample, but the saninizer does:

```
> clang++ -O3 -fsanitize=address,undefined memcmp.cpp && ./a.out
memcmp.cpp:8:5: runtime error: null pointer passed as argument 1, which is declared to never be null
/usr/include/string.h:65:33: note: nonnull attribute specified here
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior memcmp.cpp:8:5 in 
memcmp.cpp:8:5: runtime error: null pointer passed as argument 2, which is declared to never be null
/usr/include/string.h:65:33: note: nonnull attribute specified here
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior memcmp.cpp:8:5 in 
```

BTW, my string.h header contains the following memcmp declaration, directly specifying that pointers must not be null:

```c++
/* Compare N bytes of S1 and S2.  */
extern int memcmp (const void *__s1, const void *__s2, size_t __n)
     __THROW __attribute_pure__ __nonnull ((1, 2));
```

> I use valgrind-3.17.0 with no option.

I have the same valgrind version. If you directly use valgrind (like `valgrind ctest`), please try to use `ctest -T memcheck` as mentioned above - these commands behave differently on my machine. Unfortunately, I can't tell you what is the difference because I do not use valgrind very often. The fact is that the problem was found when I was experimenting with CI builds (see https://github.com/kovdan01/msgpack-c/runs/2697426801). The build logs are already expired, but you can see that builds with api version 1 fail.

redboltz replied:
Do you know how to execute only `raw` using `ctest -T memcheck` ?

redboltz replied:
`ctest -T memcheck` doesn't report error.
But as you mentioned address sanitizer with undefined option reports error.
I accept the code change.
Thanks.

kovdan01 replied:
Yes, use this:
```
ctest -T memcheck -R raw
```

kovdan01 replied:
> I accept the code change.
Thanks.

Ok, thank you. I assume that I can resolve the conversation, deel free to re-open it if needed.

redboltz commented on include/msgpack/v1/adaptor/string.hpp:
```diff
@@ -30,10 +30,16 @@ struct convert<std::string> {
+            if (o.via.bin.size == 0)
+                v.clear();
+            else
+                v.assign(o.via.bin.ptr, o.via.bin.size);
```
This is similar code to `memcmp()` we discussed.
But I think it is not requred.

In this case, both valgrind and sanitizer don't report error.

redboltz replied:
libc++ implementation is 
https://github.com/llvm/llvm-project/blob/main/libcxx/include/string#L1909-L1919

libstdc++ implementation is
https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/basic_string.h#L1440-L1445
https://github.com/gcc-mirror/gcc/blob/16e2427f50c208dfe07d07f18009969502c25dc8/libstdc%2B%2B-v3/include/debug/debug.h#L122-L123
https://github.com/gcc-mirror/gcc/blob/16e2427f50c208dfe07d07f18009969502c25dc8/libstdc%2B%2B-v3/include/debug/macros.h#L462-L463

redboltz replied:
I will update the msgpack code and check it using cmake -T memcheck soon.

redboltz replied:
I've tested both sanitizer undefined checking and valgrind for all tests on my local environment, and got no error.

kovdan01 replied:
I tried to revert the change, and now CI build fails: https://github.com/kovdan01/msgpack-c/pull/9/checks?check_run_id=3450115223.

It is also reproduced on my local environment (`ctest -T memcheck -R user_class`). More verbose output:

```
==16998== Memcheck, a memory error detector
==16998== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==16998== Using Valgrind-3.17.0 and LibVEX; rerun with -h for copyright info
==16998== Command: /home/kovdan01/Projects/msgpack-c/build-problem/test/user_class
==16998== Parent PID: 16997
==16998==
==16998== Conditional jump or move depends on uninitialised value(s)
==16998==    at 0x4AB008B: _M_disjunct (basic_string.h:340)
==16998==    by 0x4AB008B: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_replace(unsigned long, unsigned long, char const*, unsigned long) (basic_string.tcc:462)
==16998==    by 0x14D212: msgpack::v1::adaptor::convert<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, void>::operator()(msgpack::v1::object const&, std::__cxx11::basic_stri>
==16998==    by 0x15B387: msgpack::v1::enable_if<!msgpack::v1::is_array<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::value, msgpack::v1::object const&>::type msgpack::v1>
==16998==    by 0x158AD0: msgpack::v1::enable_if<(!msgpack::v1::is_array<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::value)&&(!msgpack::v1::is_pointer<std::__cxx11::bas>
==16998==    by 0x155BAB: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > msgpack::v1::object::as<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>
==16998==    by 0x14EA1C: msgpack::v1::adaptor::convert<test_non_intrusive, void>::operator()(msgpack::v1::object const&, test_non_intrusive&) const (user_class.cpp:503)
==16998==    by 0x15CC4D: msgpack::v1::enable_if<!msgpack::v1::is_array<test_non_intrusive>::value, msgpack::v1::object const&>::type msgpack::v1::operator>><test_non_intrusive>(msgpack::v1::object const&, t>
==16998==    by 0x15A4EE: msgpack::v1::enable_if<(!msgpack::v1::is_array<test_non_intrusive>::value)&&(!msgpack::v1::is_pointer<test_non_intrusive>::value), test_non_intrusive&>::type msgpack::v1::object::co>
==16998==    by 0x155CE7: test_non_intrusive msgpack::v1::object::as<test_non_intrusive>() const (object.hpp:1114)
==16998==    by 0x145979: test_non_intrusive_::test_method() (user_class.cpp:542)
==16998==    by 0x1455A3: test_non_intrusive__invoker() (user_class.cpp:534)
==16998==    by 0x15D439: boost::detail::function::void_function_invoker0<void (*)(), void>::invoke(boost::detail::function::function_buffer&) (function_template.hpp:117)
==16998==
==16998==
==16998== HEAP SUMMARY:
==16998==     in use at exit: 0 bytes in 0 blocks
==16998==   total heap usage: 31,446 allocs, 31,446 frees, 58,095,399 bytes allocated
==16998==
==16998== All heap blocks were freed -- no leaks are possible
==16998==
==16998== Use --track-origins=yes to see where uninitialised values come from
==16998== For lists of detected and suppressed errors, rerun with: -s
==16998== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)

```

I am unable to detect the problem with sanitizers, looks like it is only detected with valgrind.

redboltz replied:
It is very weird. I don't understand that why that special case is failed.
I want to know why it happens.
How about minimal example like https://github.com/msgpack/msgpack-c/pull/951#discussion_r697843338 with the same compiler setting and the same valgrind setting?
I will also check it when I have a time.

kovdan01 replied:
OK, I'll try to create a minimal example that represents the problem. I'll inform you about the results.

redboltz replied:
It is not minimal enough but I found one reproduce code:

```cpp
#include <cassert>
#include <string>

struct object;

struct object_str {
    uint32_t size;
    const char* ptr;
};

struct object_bin {
    uint32_t size;
    const char* ptr;
};

struct object_array {
    uint32_t size;
    object* ptr;
};

struct object {
    union {
        object_str str;
        object_bin bin;
        object_array array;
    } via;
};

inline void convert(object const& o, std::string& v) {
    v.assign(o.via.str.ptr, o.via.str.size);
}

std::string  as(object const& o) {
    std::string v;
    convert(o, v);
    return v;
}

struct my {
    void set_name(std::string const& name) {
        this->name = name;
    }
    std::string name;
};

int main() {
    char const* p1 = (char*)(0x1);
    object o;
    o.via.str.size = 0;
    // o.via.str.ptr = nullptr;
    my m;
    m.set_name(as(o));
}
```

If I uncomment the null initialize, then the report is disappeared.

redboltz replied:
The reporting happens only if MSGPACK_API_VERSION=1. It is a important point.

### API_VERSION 1 unpack

https://github.com/msgpack/msgpack-c/blob/be4d971c62798eb59f8455dc77a4529748bcd08f/include/msgpack/v1/unpack.hpp#L167-L197

### API_VERION >= 2 unpack

https://github.com/msgpack/msgpack-c/blob/be4d971c62798eb59f8455dc77a4529748bcd08f/include/msgpack/v2/create_object_visitor.hpp#L110-L157

### Hypothesis
When string size is 0, on API_VERSION 1 unpack's `o.via.str.ptr` is uninitialized but API_VERSION >=2 unpack's `o.via.str.ptr` is initialized as nullptr.

I think that it is essential reason.

redboltz replied:
Could you test the following fix?
If it works, please add the fix to this PR. And remove size check from assign().

```diff
diff --git a/include/msgpack/v1/unpack.hpp b/include/msgpack/v1/unpack.hpp
index 4f5dab47..fefee66d 100644
--- a/include/msgpack/v1/unpack.hpp
+++ b/include/msgpack/v1/unpack.hpp
@@ -177,6 +177,9 @@ inline void unpack_str(unpack_user& u, const char* p, uint32_t l, msgpack::objec
         std::memcpy(tmp, p, l);
         o.via.str.ptr = tmp;
     }
+    else {
+        o.via.str.ptr = MSGPACK_NULLPTR;
+    }
     o.via.str.size = l;
 }

@@ -193,6 +196,9 @@ inline void unpack_bin(unpack_user& u, const char* p, uint32_t l, msgpack::objec
         std::memcpy(tmp, p, l);
         o.via.bin.ptr = tmp;
     }
+    else {
+        o.via.bin.ptr = MSGPACK_NULLPTR;
+    }
     o.via.bin.size = l;
 }
```

kovdan01 replied:
The fix works (both locally and on CI), thanks! You are right – the root cause was in uninitialized ptr value. Updated the PR (added the appropriate fix & deleted unneeded size check in assign code).

redboltz commented:
@kovdan01 Thank you for updating the PR. 
Finally, all issues have been solved. I appreciate your great effort!

kovdan01 commented:
@redboltz Thank you for reviewing such a huge PR!
